alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+                auto v = meta[*expiration_column]->type->deserialize(*cell);
+                bool expired = false;
+                // FIXME: don't recalculate "now" all the time
+                auto now = gc_clock::now();
+                if (member) {
+                    // In this case, the expiration-time attribute we're
+                    // looking for is a member in a map, saved serialized
+                    // into bytes using Alternator's serialization (basically
+                    // a JSON serialized into bytes)
+                    // FIXME: is it possible to find a specific member of a map
+                    // without iterating through it like we do here and compare
+                    // the key?
+                    for (const auto& entry : value_cast<map_type_impl::native_type>(v)) {
+                        std::string attr_name = value_cast<sstring>(entry.first);
+                        if (value_cast<sstring>(entry.first) == *member) {
+                            bytes value = value_cast<bytes>(entry.second);
+                            rjson::value json = deserialize_item(value);
+                            expired = is_expired(json, now);
+                            break;
+                        }
+                    }
+                } else {
+                    // For a real column to contain an expiration time, it
+                    // must be a numeric type.
+                    // FIXME: Currently we only support decimal_type (which is
+                    // what Alternator uses), but other numeric types can be
+                    // supported as well to make this feature more useful in CQL.
+                    // Note that kind::decimal is also checked above.
+                    big_decimal n = value_cast<big_decimal>(v);
+                    expired = is_expired(n, now);
+                }
+                if (expired) {
+                    // FIXME: maybe don't recalculate new_timestamp() all the time
+                    auto ts = api::new_timestamp();
+                    co_await expire_item(_proxy, *query_state_ptr, row, pk_columns, ck_columns, s, ts);
+                }
+            }
+            // FIXME: continue with next page if needed. Also once in a while, persist our position.
+            //if (!p->is_exhausted()) {
+            //    ... persist p->state();
+            //    continue to fetch more pages.
+            //}
+        }
+        // FIXME: replace this silly 1-second sleep by something smarter.
+        co_await seastar::sleep(std::chrono::seconds(1));
+    }
+    co_return;
+}
+
+future<> expiration_service::start() {
+    // Called by main() on each shard to start the expiration-service
+    // thread. Just runs run() in the background.
+    // FIXME: we need to have a way to stop the thread we started (abort_source)
+    if (_db.features().cluster_supports_alternator_ttl()) {
+        (void) run().handle_exception([] (std::exception_ptr ep) {
```

Comment:
You should capture this future and await it in stop().

Reply:
Right. I left FIXMEs (as in the case for stopping) and need to start doing them either now, or after committing some initial version so I don't need to work on maintaining a large patch set and you don't need to review it over and over.

By the way, I just added paging to the code so at least that FIXME is done and now expiration works even on a larger table (still without any reasonable pacing).